app/github_repo_fetcher.rb

```diff
@@ -32,13 +32,22 @@ def docs(app_name)
           title: title,
           markdown: contents,
           source_url: doc.html_url,
+          latest_commit: latest_commit(app_name, doc.path),
         }
       end
     rescue Octokit::NotFound
       nil
     end
   end
 
+  def latest_commit(app_name, path)
+    latest_commit = client.commits("alphagov/#{app_name}", repo(app_name).default_branch, path: path).first
```

Comment:
1 request for every doc we pull in. Agreed it's not ideal. Can't really think of any alternatives at present 🤔

Reply:
That's OK - my [one suggestion](https://github.com/alphagov/govuk-developer-docs/pull/2784#discussion_r503754490) on this didn't pan out. We should still make sure it's not too significant - do you have any rough stats on the before/after performance for populating the cache?

Reply:
My benchmark command is `rm -rf .cache && GITHUB_TOKEN=$(cat ~/github_token.txt) NO_CONTRACTS=true bundle exec middleman build`.

| Branch | Build time | Requests to GitHub API |
|--------|-----------|-------------------------|
| master | 1m 48s     | 20 |
| last-updated | 2m 02s | 115 |
| last-updated (with `per_page: 1` suggestion) | 3m 20s | 770 |

Using a `per_page` of 1 exponentially increases the number of API requests, for some reason.
Build time for this branch versus main branch is comparable, though obviously quite a few more API requests.
It seems manageable for now, but will need to revisit if/when we start importing _all_ repos '`docs/`' folder.

Reply:
😆 - it seems to be because of [the auto-pagination behaviour](https://github.com/octokit/octokit.rb/blob/519512aed8bbbc781ad1c1743c162aa6dbcb4862/lib/octokit/connection.rb#L86).

It would be good if we could switch it off, as apart from the size of the response, it will cut down on extra requests. (Although it does seem unlikely that a given doc would have over 100 commits against it!)

Reply:
I've tried again with `per_page=1` and `auto_paginate=false` and ended up with about a 3 minute build time and 191 requests. Again, for reasons I'm not entirely sure of, we're seeing multiple requests - here's a snippet below:

<details>
<summary>Summary</summary>
```
I, [2020-10-22T13:35:49.639699 #91115]  INFO -- request: GET https://raw.githubusercontent.com/alphagov/publishing-api/master/docs/dependency-resolution.md
I, [2020-10-22T13:35:49.639779 #91115]  INFO -- request: User-Agent: "Faraday v1.1.0"
I, [2020-10-22T13:35:49.914402 #91115]  INFO -- response: Status 200
I, [2020-10-22T13:35:49.914556 #91115]  INFO -- response: connection: "keep-alive"
content-length: "1782"
content-type: "text/plain; charset=utf-8"
cache-control: "max-age=300"
content-security-policy: "default-src 'none'; style-src 'unsafe-inline'; sandbox"
etag: "W/\"56d0bb8508819f5ac533cca6b71a30364839d415f3268b910d16962a518d3a4a\""
strict-transport-security: "max-age=31536000"
x-content-type-options: "nosniff"
x-frame-options: "deny"
x-xss-protection: "1; mode=block"
via: "1.1 varnish (Varnish/6.0), 1.1 varnish"
x-github-request-id: "2496:170E:102AD66:113EF08:5F9178F2"
accept-ranges: "bytes"
date: "Thu, 22 Oct 2020 12:35:49 GMT"
x-served-by: "cache-lhr7377-LHR"
x-cache: "MISS, MISS"
x-cache-hits: "0, 0"
x-timer: "S1603370150.721738,VS0,VE180"
vary: "Authorization,Accept-Encoding, Accept-Encoding"
access-control-allow-origin: "*"
x-fastly-request-id: "4e0977f12fab3563805a02906182d42d067c1dfa"
expires: "Thu, 22 Oct 2020 12:40:49 GMT"
source-age: "0"
I, [2020-10-22T13:35:49.919305 #91115]  INFO -- request: GET https://api.github.com/repos/alphagov/publishing-api/commits?auto_paginate=false&path=docs%2Fdependency-resolution.md&per_page=1&sha=master
I, [2020-10-22T13:35:49.919398 #91115]  INFO -- request: Accept: "application/vnd.github.mercy-preview+json"
User-Agent: "Octokit Ruby Gem 4.19.0"
Content-Type: "application/json"
Authorization: "token redacted"
I, [2020-10-22T13:35:50.270074 #91115]  INFO -- response: Status 200
I, [2020-10-22T13:35:50.270224 #91115]  INFO -- response: date: "Thu, 22 Oct 2020 12:35:50 GMT"
content-type: "application/json; charset=utf-8"
transfer-encoding: "chunked"
server: "GitHub.com"
status: "200 OK"
cache-control: "private, max-age=60, s-maxage=60"
vary: "Accept, Authorization, Cookie, X-GitHub-OTP, Accept-Encoding, Accept, X-Requested-With, Accept-Encoding"
etag: "W/\"37e9bcf952753b113c6c03075e3a5df284b02aa0493a83d4088444ffc6b15aa2\""
last-modified: "Tue, 09 Jun 2020 15:20:16 GMT"
x-oauth-scopes: "admin:org, admin:org_hook, admin:public_key, admin:repo_hook, repo, workflow"
x-accepted-oauth-scopes: ""
x-github-media-type: "github.mercy-preview; format=json"
link: "<https://api.github.com/repositories/27546098/commits?auto_paginate=false&path=docs%2Fdependency-resolution.md&per_page=1&sha=master&page=2>; rel=\"next\", <https://api.github.com/repositories/27546098/commits?auto_paginate=false&path=docs%2Fdependency-resolution.md&per_page=1&sha=master&page=2>; rel=\"last\""
x-ratelimit-limit: "5000"
x-ratelimit-remaining: "4741"
x-ratelimit-reset: "1603372753"
x-ratelimit-used: "259"
access-control-expose-headers: "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset"
access-control-allow-origin: "*"
strict-transport-security: "max-age=31536000; includeSubdomains; preload"
x-frame-options: "deny"
x-content-type-options: "nosniff"
x-xss-protection: "1; mode=block"
referrer-policy: "origin-when-cross-origin, strict-origin-when-cross-origin"
content-security-policy: "default-src 'none'"
x-github-request-id: "F89B:7C68:22AA8BD7:280B1A9C:5F917CA5"
I, [2020-10-22T13:35:50.272136 #91115]  INFO -- request: GET https://api.github.com/repositories/27546098/commits?auto_paginate=false&page=2&path=docs%2Fdependency-resolution.md&per_page=1&sha=master
I, [2020-10-22T13:35:50.272205 #91115]  INFO -- request: Accept: "application/vnd.github.mercy-preview+json"
User-Agent: "Octokit Ruby Gem 4.19.0"
Content-Type: "application/json"
Authorization: "token redacted"
I, [2020-10-22T13:35:50.577923 #91115]  INFO -- response: Status 200
I, [2020-10-22T13:35:50.578082 #91115]  INFO -- response: date: "Thu, 22 Oct 2020 12:35:50 GMT"
content-type: "application/json; charset=utf-8"
transfer-encoding: "chunked"
server: "GitHub.com"
status: "200 OK"
cache-control: "private, max-age=60, s-maxage=60"
vary: "Accept, Authorization, Cookie, X-GitHub-OTP, Accept-Encoding, Accept, X-Requested-With, Accept-Encoding"
etag: "W/\"58828caa022c22b94fd7b40e99113d0bcf9c0f23c94e3f6557f34a1695381527\""
last-modified: "Tue, 09 Jun 2020 15:17:06 GMT"
x-oauth-scopes: "admin:org, admin:org_hook, admin:public_key, admin:repo_hook, repo, workflow"
x-accepted-oauth-scopes: ""
x-github-media-type: "github.mercy-preview; format=json"
link: "<https://api.github.com/repositories/27546098/commits?auto_paginate=false&page=3&path=docs%2Fdependency-resolution.md&per_page=1&sha=master>; rel=\"next\", <https://api.github.com/repositories/27546098/commits?auto_paginate=false&page=2&path=docs%2Fdependency-resolution.md&per_page=1&sha=master>; rel=\"last\", <https://api.github.com/repositories/27546098/commits?auto_paginate=false&page=1&path=docs%2Fdependency-resolution.md&per_page=1&sha=master>; rel=\"first\", <https://api.github.com/repositories/27546098/commits?auto_paginate=false&page=1&path=docs%2Fdependency-resolution.md&per_page=1&sha=master>; rel=\"prev\""
x-ratelimit-limit: "5000"
x-ratelimit-remaining: "4740"
x-ratelimit-reset: "1603372753"
x-ratelimit-used: "260"
access-control-expose-headers: "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset"
access-control-allow-origin: "*"
strict-transport-security: "max-age=31536000; includeSubdomains; preload"
x-frame-options: "deny"
x-content-type-options: "nosniff"
x-xss-protection: "1; mode=block"
referrer-policy: "origin-when-cross-origin, strict-origin-when-cross-origin"
content-security-policy: "default-src 'none'"
x-github-request-id: "F89C:179A:8835ACC:9C2C0AB:5F917CA6"
I, [2020-10-22T13:35:50.579847 #91115]  INFO -- request: GET https://api.github.com/repositories/27546098/commits?auto_paginate=false&page=3&path=docs%2Fdependency-resolution.md&per_page=1&sha=master
I, [2020-10-22T13:35:50.579902 #91115]  INFO -- request: Accept: "application/vnd.github.mercy-preview+json"
User-Agent: "Octokit Ruby Gem 4.19.0"
Content-Type: "application/json"
Authorization: "token redacted"
I, [2020-10-22T13:35:50.837166 #91115]  INFO -- response: Status 200
I, [2020-10-22T13:35:50.837331 #91115]  INFO -- response: date: "Thu, 22 Oct 2020 12:35:50 GMT"
content-type: "application/json; charset=utf-8"
content-length: "2"
server: "GitHub.com"
status: "200 OK"
cache-control: "private, max-age=60, s-maxage=60"
vary: "Accept, Authorization, Cookie, X-GitHub-OTP, Accept-Encoding, Accept, X-Requested-With, Accept-Encoding"
etag: "\"58ab382c69a71f1296eaa9054cefc6ae629249925f5aef3b66b30557d44a1872\""
x-oauth-scopes: "admin:org, admin:org_hook, admin:public_key, admin:repo_hook, repo, workflow"
x-accepted-oauth-scopes: ""
x-github-media-type: "github.mercy-preview; format=json"
link: "<https://api.github.com/repositories/27546098/commits?auto_paginate=false&page=1&path=docs%2Fdependency-resolution.md&per_page=1&sha=master>; rel=\"first\", <https://api.github.com/repositories/27546098/commits?auto_paginate=false&page=2&path=docs%2Fdependency-resolution.md&per_page=1&sha=master>; rel=\"prev\""
x-ratelimit-limit: "5000"
x-ratelimit-remaining: "4739"
x-ratelimit-reset: "1603372753"
x-ratelimit-used: "261"
access-control-expose-headers: "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset"
access-control-allow-origin: "*"
strict-transport-security: "max-age=31536000; includeSubdomains; preload"
x-frame-options: "deny"
```
</details>

Here's the code I used:

```ruby
    latest_commit = client.commits("alphagov/#{app_name}", repo(app_name).default_branch, path: path, per_page: 1, auto_paginate: false).first
```

It seems as though the Octokit client basically ignores the `auto_paginate` flag (although I'm not sure why it has resulted in 191 requests, vs 770 yesterday). The documentation around this is poor. It's also worrying that the `commits` method [appears to be deprecated](https://www.rubydoc.info/github/pengwynn/octokit/Octokit/Client/Commits#commits-instance_method).

So far it's still looking like the version without `per_page` and `auto_paginate` is the best to go with.

Reply:
Yeah, I noticed that `auto_paginate` didn't seem to be configurable! While the method in Octokit may be deprecated (that page 404s, but I'm assuming you're right), it's [still in the GitHub API](https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#list-commits), so hopefully nothing to worry about there.